pkg/report/coverageprofile.go

```diff
@@ -0,0 +1,306 @@
+	})
+
+	violationsMap := make(map[int]bool)
+	// NumStmt indicates the number of statements in a code block, it does not means the line, because a statement may have several lines,
+	// which means that the value of NumStmt is less or equal tothe total numbers of the code block.
+	for _, b := range blocks {
+		total += int64(b.NumStmt)
+		if b.Count > 0 {
+			covered += int64(b.NumStmt)
+		} else {
+			for i := b.StartLine; i <= b.EndLine; i++ {
+				violationsMap[i] = true
+			}
+		}
+	}
+
+	// it actually should not happen
+	if total == 0 {
+		return nil
+	}
+
+	violationLines := sortLines(violationsMap)
+
+	coveredPercent := int(float64(covered) / float64(total) * 100)
+
+	coverageProfile := &CoverageProfile{
+		FileName:            change.FileName,
+		TotalLines:          int(total),
+		CoveredLines:        int(covered),
+		CoveragePercent:     coveredPercent,
+		TotalViolationLines: violationLines,
+		ViolationSections: []*ViolationSection{
+			{
+				ViolationLines: violationLines,
+				StartLine:      change.Sections[0].StartLine,
+				EndLine:        change.Sections[0].EndLine,
+				Contents:       change.Sections[0].Contents,
+			},
+		},
+	}
+
+	return coverageProfile
+}
+
+// generateCoverageProfileWithModifyMode generates for modify file
+func generateCoverageProfileWithModifyMode(profile *cover.Profile, change *gittool.Change) *CoverageProfile {
+
+	blocks := profile.Blocks
+	sort.Slice(blocks, func(i, j int) bool {
+		return blocks[i].StartLine < blocks[j].StartLine
+	})
+
+	var total, covered int64
+	var totalViolationLines []int
+	var violationSections []*ViolationSection
+
+	// for each section contents, find each line in profile block and judge it
+	for _, section := range change.Sections {
+
+		var violationLines []int
+		for lineNo := section.StartLine; lineNo <= section.EndLine; lineNo++ {
+			block := binarySeachForProfileBlock(blocks, 0, len(blocks)-1, lineNo)
+			if block == nil {
+				continue
+			}
+
+			// check line?
+
+			total++
+			if block.Count > 0 {
+				covered++
+			} else {
+				violationLines = append(violationLines, lineNo)
+			}
+
+		}
+
+		if len(violationLines) != 0 {
+			violationSections = append(violationSections, &ViolationSection{
+				StartLine:      section.StartLine,
+				EndLine:        section.EndLine,
+				Contents:       section.Contents,
+				ViolationLines: violationLines,
+			})
+
+			totalViolationLines = append(totalViolationLines, violationLines...)
+		}
+
+	}
+
+	if total == 0 {
+		return nil
+	}
+
+	coveredPercent := int(float64(covered) / float64(total) * 100)
+
+	return &CoverageProfile{
+		FileName:            change.FileName,
+		TotalLines:          int(total),
+		CoveredLines:        int(covered),
+		CoveragePercent:     coveredPercent,
+		TotalViolationLines: totalViolationLines,
+		ViolationSections:   violationSections,
+	}
+}
+
+func binarySeachForProfileBlock(blocks []cover.ProfileBlock, left int, right int, lineNo int) *cover.ProfileBlock {
```

Comment:
nit: we can make use of `sort.Search` to perform the binary search

https://pkg.go.dev/sort#Search

Reply:
`sort.Seach` returns the index and need to check the index and the key element. In current situation, it looks like it's not applicable for this case, because it returns a `interface{}`.

Reply:
```
idx := sort.Search()
return &blocks[idx]
```

Why this is inapplicable to this case?

BTW, your binary search and the sorting implementation is incorrect as

```
sort.Slice(blocks, func(i, j int) bool {
		return blocks[i].StartLine < blocks[j].StartLine
	})
```

Is sorting by start line only, but for cases like `a.startline == b.startline`, the order is incorrect as we don't check the endline here. I would suggest we create a slice type for the cover.ProfileBlock and implement the `sort.Interface`. Then we can leverage the stdlib to do the heavylifting.

Reply:
take advantage of sort.Interface now.